File: backend/app/schemas/batch.py
"""
Batch Processing Schemas

Pydantic models for batch upload, job status, and results.
"""
from typing import Optional, List, Dict, Any, Literal
from pydantic import BaseModel, Field


class BatchUploadResponse(BaseModel):
    """Response from batch file upload."""

    job_id: str = Field(..., description="Unique job identifier for tracking")
    status: str = Field(default="pending", description="Initial job status")
    total_molecules: int = Field(..., description="Number of molecules detected in file")
    message: str = Field(default="Job submitted successfully")


class BatchJobStatus(BaseModel):
    """Current status of a batch job."""

    job_id: str
    status: Literal["pending", "processing", "complete", "failed", "cancelled"]
    progress: int = Field(..., ge=0, le=100, description="Progress percentage")
    processed: int = Field(..., description="Number of molecules processed")
    total: int = Field(..., description="Total number of molecules")
    eta_seconds: Optional[int] = Field(None, description="Estimated time remaining in seconds")
    error_message: Optional[str] = Field(None, description="Error message if failed")


class BatchResultItem(BaseModel):
    """Single molecule result in batch."""

    smiles: str
    name: Optional[str] = None
    index: int
    status: Literal["success", "error"]
    error: Optional[str] = None
    validation: Optional[Dict[str, Any]] = Field(
        None, description="Validation results with overall_score and issues"
    )
    alerts: Optional[Dict[str, Any]] = Field(
        None, description="Structural alert screening results"
    )
    scoring: Optional[Dict[str, Any]] = Field(
        None, description="ML-readiness scoring results"
    )


class BatchStatistics(BaseModel):
    """Aggregate statistics for a batch job."""

    total: int = Field(..., description="Total molecules processed")
    successful: int = Field(..., description="Molecules processed successfully")
    errors: int = Field(..., description="Molecules with errors")
    avg_validation_score: Optional[float] = Field(
        None, description="Average validation score"
    )
    avg_ml_readiness_score: Optional[float] = Field(
        None, description="Average ML-readiness score"
    )
    score_distribution: Dict[str, int] = Field(
        default_factory=dict,
        description="Score distribution buckets (excellent, good, moderate, poor)",
    )
    alert_summary: Dict[str, int] = Field(
        default_factory=dict, description="Alert counts by catalog type"
    )
    processing_time_seconds: Optional[float] = Field(
        None, description="Total processing time in seconds"
    )


class BatchResultsResponse(BaseModel):
    """Paginated batch results response."""

    job_id: str
    status: str
    statistics: Optional[BatchStatistics] = None
    results: List[BatchResultItem] = Field(default_factory=list)
    page: int = Field(..., ge=1)
    page_size: int = Field(..., ge=1, le=100)
    total_results: int
    total_pages: int


class CSVColumnsResponse(BaseModel):
    """Response with detected CSV columns for SMILES selection."""

    columns: List[str] = Field(..., description="Available column names")
    suggested_smiles: Optional[str] = Field(
        None, description="Suggested SMILES column based on name"
    )
    row_count_estimate: int = Field(..., description="Estimated number of rows")


class BatchProgressMessage(BaseModel):
    """WebSocket progress message format."""

    job_id: str
    status: str
    progress: int = Field(..., ge=0, le=100)
    processed: int
    total: int
    eta_seconds: Optional[int] = None
    error_message: Optional[str] = None

File: frontend/src/types/batch.ts
/**
 * Batch Processing Types
 *
 * TypeScript interfaces for batch upload, progress tracking, and results display.
 */

/**
 * Progress update from WebSocket
 */
export interface BatchProgress {
  job_id: string;
  status: 'pending' | 'processing' | 'complete' | 'failed' | 'cancelled';
  progress: number; // 0-100
  processed: number;
  total: number;
  eta_seconds: number | null;
  error_message?: string | null;
}

/**
 * Single molecule result in batch
 */
export interface BatchResult {
  smiles: string;
  name: string | null;
  index: number;
  status: 'success' | 'error';
  error: string | null;
  validation: {
    overall_score: number;
    issues?: Array<{
      check_name: string;
      passed: boolean;
      severity: string;
      message: string;
    }>;
    error?: string;
  } | null;
  alerts: {
    has_alerts: boolean;
    alert_count: number;
    alerts: Array<{
      catalog: string;
      rule_name: string;
      severity: string;
    }>;
    error?: string;
  } | null;
  scoring: {
    ml_readiness: {
      score: number;
      interpretation: string;
    };
    error?: string;
  } | null;
}

/**
 * Aggregate statistics for batch job
 */
export interface BatchStatistics {
  total: number;
  successful: number;
  errors: number;
  avg_validation_score: number | null;
  avg_ml_readiness_score: number | null;
  score_distribution: {
    excellent: number;
    good: number;
    moderate: number;
    poor: number;
  };
  alert_summary: Record<string, number>;
  processing_time_seconds: number | null;
}

/**
 * Response from batch upload endpoint
 */
export interface BatchUploadResponse {
  job_id: string;
  status: string;
  total_molecules: number;
  message: string;
}

/**
 * Response from batch results endpoint
 */
export interface BatchResultsResponse {
  job_id: string;
  status: string;
  statistics: BatchStatistics | null;
  results: BatchResult[];
  page: number;
  page_size: number;
  total_results: number;
  total_pages: number;
}

/**
 * Response from column detection endpoint
 */
export interface CSVColumnsResponse {
  columns: string[];
  suggested_smiles: string | null;
  row_count_estimate: number;
}

/**
 * Filters for results query
 */
export interface BatchResultsFilters {
  status_filter?: 'success' | 'error';
  min_score?: number;
  max_score?: number;
}

/**
 * Batch validation page state
 */
export type BatchPageState = 'upload' | 'processing' | 'results';

/**
 * Export format options (matches backend ExportFormat enum)
 */
export type ExportFormat = 'csv' | 'excel' | 'sdf' | 'json' | 'pdf';
